plat/kvm/efi.c

```diff
@@ -0,0 +1,342 @@
+		mrd->type = UKPLAT_MEMRT_RESERVED;
+		mrd->flags = UKPLAT_MEMRF_READ | UKPLAT_MEMRF_WRITE |
+			     UKPLAT_MEMRF_MAP;
+
+		break;
+	case UK_EFI_RUNTIME_SERVICES_CODE:
+	case UK_EFI_RUNTIME_SERVICES_DATA:
+		/* Already added through uk_efi_rt_md_to_bi_mrds() */
+		return -EEXIST;
+	case UK_EFI_LOADER_CODE:
+	case UK_EFI_LOADER_DATA:
+		/* Already added through mkbootinfo.py and relocated through
+		 * do_ukreloc
+		 */
+		return -EEXIST;
+	case UK_EFI_BOOT_SERVICES_CODE:
+	case UK_EFI_BOOT_SERVICES_DATA:
+	case UK_EFI_CONVENTIONAL_MEMORY:
+		/* These are freed after ExitBootServices is called */
+		mrd->type = UKPLAT_MEMRT_FREE;
+
+		mrd->flags = UKPLAT_MEMRF_READ | UKPLAT_MEMRF_WRITE;
+
+		break;
+	}
+
+	/* Ignore zero-page */
+	start = MAX(md->physical_start, __PAGE_SIZE);
+	end = md->physical_start + md->number_of_pages * UK_EFI_PAGE_SIZE;
+	if (unlikely(end <= start || end - start < __PAGE_SIZE))
+		return -ENOMEM;
+
+	mrd->pbase = start;
+	mrd->vbase = start;
+	mrd->len = end - start;
+
+	return 0;
+}
+
+static void uk_efi_get_mmap(uk_efi_mem_desc_t **map, uk_efi_uintn_t *map_sz,
+			    uk_efi_uintn_t *desc_sz)
+{
+	uk_efi_status_t status;
+	__u32 desc_ver;
+
+	/* As the UEFI Spec says:
+	 * If the MemoryMap buffer is too small, the EFI_BUFFER_TOO_SMALL
+	 * error code is returned and the MemoryMapSize value contains the
+	 * size of the buffer needed to contain the current memory map. The
+	 * actual size of the buffer allocated for the consequent call to
+	 * GetMemoryMap() should be bigger then the value returned in
+	 * MemoryMapSize, since allocation of the new buffer may potentially
+	 * increase memory map size.
+	 */
+	*map_sz = 0;  /* force EFI_BUFFER_TOO_SMALL */
+	*map = NULL;
+	status = uk_efi_bs->get_memory_map(map_sz, *map, &uk_efi_map_key,
+					   desc_sz, &desc_ver);
+	if (unlikely(status != UK_EFI_BUFFER_TOO_SMALL))
+		uk_efi_crash("Failed to call initial dummy get_memory_map\n");
+
+	/* Make sure the actual allocated buffer is bigger */
+	*map_sz += *desc_sz * UK_EFI_SURPLUS_MEM_DESC_COUNT;
+	status = uk_efi_bs->allocate_pool(UK_EFI_LOADER_DATA, *map_sz,
+					  (void **)map);
+	if (unlikely(status != UK_EFI_SUCCESS))
+		uk_efi_crash("Failed to allocate memory for map\n");
+
+	/* Now we call it for real */
+	status = uk_efi_bs->get_memory_map(map_sz, *map, &uk_efi_map_key,
+					   desc_sz, &desc_ver);
+	if unlikely((status != UK_EFI_SUCCESS))
+		uk_efi_crash("Failed to get memory map\n");
+}
+
+/* Runtime Services memory regions in the Memory Attribute Table have a higher
+ * granularity regarding sizes and permissions: the ones resulted from
+ * GetMemoryMap only differentiate between Runtime Services Data/Code, while
+ * the MAT also differentiates between permissions of the Runtime Services'
+ * PE sections (Runtime Services can basically be thought of as loaded Portable
+ * Executable format drivers).
+ */
+static void uk_efi_rt_md_to_bi_mrds(struct ukplat_memregion_desc **rt_mrds,
```

Comment:
Apparently this table is optional. Just booted on a laptop that is missing the MAT. Will rework this so that if MAT is not found `uk_efi_md_to_bi_mrd` will add the memory region instead of returning `-EEXIST`